File: backend/src/modules/globalTemplateLibrary/globalTemplateLibrary.controller.js
const globalTemplateLibraryService = require('./globalTemplateLibrary.service');

async function listEquipment(req, res) {
  const rows = await globalTemplateLibraryService.listEquipmentTemplates();
  res.json(rows);
}

async function getEquipment(req, res) {
  const row = await globalTemplateLibraryService.getEquipmentTemplateById(req.params.id);
  res.json(row);
}

async function postEquipment(req, res) {
  const row = await globalTemplateLibraryService.createEquipmentTemplateFromSitePayload(req.body);
  res.status(201).json(row);
}

async function patchEquipment(req, res) {
  const row = await globalTemplateLibraryService.updateEquipmentTemplateName(req.params.id, req.body || {});
  res.json(row);
}

async function deleteEquipment(req, res) {
  await globalTemplateLibraryService.deleteEquipmentTemplate(req.params.id);
  res.status(204).end();
}

async function listGraphic(req, res) {
  const rows = await globalTemplateLibraryService.listGraphicTemplates();
  res.json(rows);
}

async function getGraphic(req, res) {
  const row = await globalTemplateLibraryService.getGraphicTemplateById(req.params.id);
  res.json(row);
}

async function postGraphic(req, res) {
  const body = req.body && typeof req.body === 'object' ? req.body : {};
  const equipmentTemplates = Array.isArray(body.equipmentTemplates) ? body.equipmentTemplates : [];
  const templatePayload = body.template && typeof body.template === 'object' ? body.template : body;
  const row = await globalTemplateLibraryService.createGraphicTemplateFromSitePayload(
    templatePayload,
    equipmentTemplates
  );
  res.status(201).json(row);
}

async function patchGraphic(req, res) {
  const row = await globalTemplateLibraryService.updateGraphicTemplateName(req.params.id, req.body || {});
  res.json(row);
}

async function deleteGraphic(req, res) {
  await globalTemplateLibraryService.deleteGraphicTemplate(req.params.id);
  res.status(204).end();
}

module.exports = {
  listEquipment,
  getEquipment,
  postEquipment,
  patchEquipment,
  deleteEquipment,
  listGraphic,
  getGraphic,
  postGraphic,
  patchGraphic,
  deleteGraphic,
};

File: server/src/main/java/com/legioncontrols/server/templates/GlobalTemplateLibraryController.java
package com.legioncontrols.server.templates;

import java.util.List;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import tools.jackson.databind.JsonNode;

/** Java equivalent of backend/src/modules/globalTemplateLibrary/globalTemplateLibrary.controller.js. */
@RestController
@RequestMapping("/api/global-template-library")
public class GlobalTemplateLibraryController {

    private final GlobalTemplateLibraryService service;

    public GlobalTemplateLibraryController(GlobalTemplateLibraryService service) {
        this.service = service;
    }

    @GetMapping("/equipment-templates")
    public List<GlobalTemplateLibraryService.EquipmentListRow> listEquipment() {
        return service.listEquipmentTemplates();
    }

    @PostMapping("/equipment-templates")
    public ResponseEntity<GlobalTemplateLibraryService.EquipmentDetail> postEquipment(@RequestBody JsonNode body) {
        return ResponseEntity.status(HttpStatus.CREATED).body(service.createEquipmentTemplateFromSitePayload(body));
    }

    @GetMapping("/equipment-templates/{id}")
    public GlobalTemplateLibraryService.EquipmentDetail getEquipment(@PathVariable String id) {
        return service.getEquipmentTemplateById(id);
    }

    @PatchMapping("/equipment-templates/{id}")
    public GlobalTemplateLibraryService.EquipmentListRow patchEquipment(@PathVariable String id, @RequestBody(required = false) JsonNode body) {
        return service.updateEquipmentTemplateName(id, body);
    }

    @DeleteMapping("/equipment-templates/{id}")
    public ResponseEntity<Void> deleteEquipment(@PathVariable String id) {
        service.deleteEquipmentTemplate(id);
        return ResponseEntity.noContent().build();
    }

    @GetMapping("/graphic-templates")
    public List<GlobalTemplateLibraryService.GraphicListRow> listGraphic() {
        return service.listGraphicTemplates();
    }

    @PostMapping("/graphic-templates")
    public ResponseEntity<GlobalTemplateLibraryService.GraphicDetail> postGraphic(@RequestBody JsonNode body) {
        JsonNode equipmentTemplates = body.has("equipmentTemplates") && body.get("equipmentTemplates").isArray() ? body.get("equipmentTemplates") : null;
        JsonNode templatePayload = body.has("template") && body.get("template").isObject() ? body.get("template") : body;
        return ResponseEntity.status(HttpStatus.CREATED).body(service.createGraphicTemplateFromSitePayload(templatePayload, equipmentTemplates));
    }

    @GetMapping("/graphic-templates/{id}")
    public GlobalTemplateLibraryService.GraphicDetail getGraphic(@PathVariable String id) {
        return service.getGraphicTemplateById(id);
    }

    @PatchMapping("/graphic-templates/{id}")
    public GlobalTemplateLibraryService.GraphicListRow patchGraphic(@PathVariable String id, @RequestBody(required = false) JsonNode body) {
        return service.updateGraphicTemplateName(id, body);
    }

    @DeleteMapping("/graphic-templates/{id}")
    public ResponseEntity<Void> deleteGraphic(@PathVariable String id) {
        service.deleteGraphicTemplate(id);
        return ResponseEntity.noContent().build();
    }
}
